jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,427 @@
+     * The entity will never be replaced.
+     * @param entity the entity to add
+     */
+    @Override
+    public void addEntity(Entity entity) {
+        LOGGER.debug("Adding/Updating the entity in ES");
+
+        //The field that has a not null value in the current Entity instance.
+        //Required, mainly for building a script for the update command.
+        String fieldToAdd = null;
+        String fieldValueToAdd = "";
+        try {
+            XContentBuilder builder = jsonBuilder()
+                    .startObject();
+
+            //Currently ignoring Graph field based indexing
+//            if (docDef.getGraphField() != null) {
+//                builder = builder.field(docDef.getGraphField(), entity.getGraph());
+//            }
+
+            for(String field: docDef.fields()) {
+                if(entity.get(field) != null) {
+                    if(entity.getLanguage() != null && !entity.getLanguage().isEmpty() && isMultilingual) {
+                        fieldToAdd = field + "_" + entity.getLanguage();
+                    } else {
+                        fieldToAdd = field;
+                    }
+
+                    fieldValueToAdd = (String) entity.get(field);
+                    builder = builder.field(fieldToAdd, Arrays.asList(fieldValueToAdd));
+                    break;
+                } else {
+                    //We are making sure that the field is at-least added to the index.
+                    //This will help us tremendously when we are appending the data later in an already indexed document.
+                    builder = builder.field(field, Collections.emptyList());
+                }
+
+            }
+
+            builder = builder.endObject();
+            IndexRequest indexRequest = new IndexRequest(INDEX_NAME, docDef.getEntityField(), entity.getId())
+                    .source(builder);
+
+            /**
+             * We are creating an upsert request here instead of a simple insert request.
+             * The reason is we want to add a document if it does not exist with the given Subject Id (URI).
+             * But if the document exists with the same Subject Id, we want to do an update to it instead of deleting it and
+             * then creating it with only the latest field values.
+             * This functionality is called Upsert functionality and more can be learned about it here:
+             * https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html#upserts
+             */
+
+            //First Search of the field exists or not
+            SearchResponse existsResponse = client.prepareSearch(INDEX_NAME)
+                    .setTypes(docDef.getEntityField())
+                    .setQuery(QueryBuilders.existsQuery(fieldToAdd))
+                    .get();
+            String script;
+            if(existsResponse != null && existsResponse.getHits() != null && existsResponse.getHits().totalHits() > 0) {
```

Comment:
Excellent comment. I will re-work on the code to make it safe and remove extra REST calls.